[SYCL][ESIMD] Add test on simd select 2d function

vasilytric opened the pull request.

vasilytric committed:
[SYCL][ESIMD] Add test on simd select 2d function

vasilytric committed:
Merge branch 'intel' into simd_select_2d

vasilytric committed:
[SYCL][ESIMD] Apply clang-format

yuriykoch commented on SYCL/ESIMD/api/functional/functions/common.hpp:
```diff
@@ -19,8 +19,7 @@ namespace esimd_test::api::functional::functions {
 
-template <int NumElems, int NumSelectedElems, int Stride, int Offset,
-          typename ContextT>
+template <int NumElems, typename ContextT>
 class TestDescription : public ITestDescription {
```
Looks like we have a same changes in #939 but from different commits
A tiny separate PR maybe?

vasilytric replied:
moving this changes into a separate PR requires applying a few changed in different PRs and when the tiny PR will be merged, then these changes should be applied in existed PRs
I suggest leave existed solution

yuriykoch replied:
Fine for me if that's more convenient for you

yuriykoch replied:
Fine for me if that's more convenient for you :)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_select_2d.hpp:
```diff
@@ -0,0 +1,286 @@
+    });
+    queue.wait_and_throw();
+
+    std::vector<size_t> selected_indexsess;
```
```suggestion
    std::vector<size_t> selected_indexes;
```
Isn't it the same misprint as we discussed in #939 and in https://github.com/intel/llvm-test-suite/pull/875#discussion_r825963443 ?

vasilytric replied:
Naming and other cases has been fixed in this commit: https://github.com/intel/llvm-test-suite/pull/908/commits/ce54fc2c173d7fdf6944b2e91334eba4dbf7f90c

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_select_2d.hpp:
```diff
@@ -0,0 +1,286 @@
+
+  template <typename DataT, int NumElems, int SizeY, int StrideY, int SizeX,
+            int StrideX, int Height, int Width>
+  static void call_esimd_function(const DataT *const ref_1,
+                                  const DataT *const ref_2, DataT *const out,
```
OK, here we go with comment from https://github.com/intel/llvm-test-suite/pull/875#discussion_r824747657 again

vasilytric replied:
Naming and other cases has been fixed in this commit: https://github.com/intel/llvm-test-suite/pull/908/commits/ce54fc2c173d7fdf6944b2e91334eba4dbf7f90c

yuriykoch replied:
What's the point in changing `ref_1` and `ref_2` into `initial_data_1` and `initial_data_2`?

You already have `initial_data` and `data_for_change` from your commit in `[functions_1d_select.hpp](https://github.com/intel/llvm-test-suite/pull/875/files#diff-295141c99d991cbb4971abc3bd754b2f17400cade3b579ac7bc36b28a1450dc3R43)`

Looks like that's the level of violating [llvm coding standards](https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly) we might easily avoid at no cost.

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_select_2d.hpp:
```diff
@@ -0,0 +1,286 @@
+           "Offset of with should be less than simd_view with size");
+
+    constexpr size_t value_for_increase_ref_data_for_fill = 50;
+    static_assert(std::numeric_limits<char>::max() >
```
One more https://github.com/intel/llvm-test-suite/pull/875#discussion_r822371818

vasilytric replied:
Naming and other cases has been fixed in this commit: https://github.com/intel/llvm-test-suite/pull/908/commits/ce54fc2c173d7fdf6944b2e91334eba4dbf7f90c

yuriykoch reviewed (commented):
Would you mind to align the patch with `functions_select_1d.hpp`? Looks like it could be done by applying the comments from #875. I might easily miss something.

vasilytric committed:
[SYCL][ESIMD] Disable sycl::half for test

Because but we can't apply new value to all selected values

vasilytric committed:
[SYCL][ESIMD] Update error message

Add offset x and offset y

vasilytric committed:
[SYCL][ESIMD] Update naming, formatting according to previous test review

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_select_2d.hpp:
```diff
@@ -0,0 +1,286 @@
+}
+
+// Helping function that lets launch test with provided tested_types.
+// IMPORTANT: do not set offset value that greater than expected number of
+// strings or element in one string in simd_view.
```
Is it possible to have `static_assert` for this condition somewhere?

vasilytric replied:
yes, they has been added in this commit: https://github.com/intel/llvm-test-suite/pull/908/commits/e81fee68457c2825f716260a4bdf088924b053c9

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_select_2d.hpp:
```diff
@@ -0,0 +1,286 @@
+    });
+    queue.wait_and_throw();
+
+    std::vector<size_t> selected_indexes;
+    // Collect the indexess that has been selected.
```
```suggestion
    std::vector<size_t> selected_indexes;
    // Collect the indexes that has been selected.
```

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/commit/cfdb4c12f261d96d9fdc262394834f02be264247#diff-2040976149a618ee4dbf449cdb7db2ebb5588203038a81b8c2468c104cee0db9R122)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_select_2d.hpp:
```diff
@@ -0,0 +1,286 @@
+    // Push the largest value to avoid the following error: can't dereference
+    // out of range vector iterator.
+    selected_indexes.push_back(std::numeric_limits<size_t>::max());
+    auto selected_indexses_ptr = selected_indexes.begin();
```
```suggestion
    auto next_selected_index = selected_indexes.begin();
```
Let's stick with the `next_selected_index` you've used for `1d select` here: https://github.com/intel/llvm-test-suite/pull/875#discussion_r824771178 - that's iterator after all.

vasilytric replied:
I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/908/commits/cfdb4c12f261d96d9fdc262394834f02be264247

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_select_2d.hpp:
```diff
@@ -0,0 +1,286 @@
+      if (i < *selected_indexses_ptr) {
+        const DataT &retrieved = result[i];
+        const DataT &expected = initial_ref_data[i];
+        if (expected != retrieved) {
+          passed = fail_test(i, expected, retrieved, data_type);
+        }
+      } else {
+        const DataT &retrieved = result[i];
+        const DataT expected = ref_data_for_change.front();
+        ref_data_for_change.erase(ref_data_for_change.begin());
+        if (expected != retrieved) {
+          passed = fail_test(i, expected, retrieved, data_type);
```
Looks like we could easily provide different error messages for each of the `fail_test` in logs. It might help to triage test failure in the future. Would you mind to add something like it's done in your other PR?

vasilytric replied:
the log error message has been updated: now, will added `value that should not be selected` or `value that should be selected` has unexpected value
I've done it in this commit: https://github.com/intel/llvm-test-suite/pull/908/commits/e57a84e8724bcb7ee46e60b16fa98b3d95403037

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_select_2d.hpp:
```diff
@@ -0,0 +1,286 @@
+        }
+      } else {
+        const DataT &retrieved = result[i];
+        const DataT expected = ref_data_for_change.front();
+        ref_data_for_change.erase(ref_data_for_change.begin());
```
Do we need to modify `ref_data_for_change` in verification loop? Maybe it's more efficient just to use an other iterator over the `ref_data_for_change`?

vasilytric replied:
the using an iterator on `ref_data_for_change` has been added in this commit: https://github.com/intel/llvm-test-suite/pull/908/commits/dcc5bfa24cc297b36642f2d9ea854fa86a8c266f

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_select_2d.hpp:
```diff
@@ -0,0 +1,286 @@
+template <tested_types TestedTypes>
+bool run_test_for_types(sycl::queue &queue) {
+  bool passed = true;
+  constexpr int desired_simd_large_size = 32;
+  constexpr int coefficient_of_division = 3;
+  constexpr int default_simd_split_value = 4;
+  constexpr int expected_height =
+      desired_simd_large_size / default_simd_split_value;
+  constexpr int expected_width = desired_simd_large_size / expected_height;
+  const int heights_large_offset =
+      expected_height - round_up_int_division(2 * expected_height, 3);
+  const int width_large_offset =
+      expected_width - round_up_int_division(2 * expected_width, 3);
```
Would you mind to comment it up a bit? It would be really helpful

vasilytric replied:
the comments has been added in this commit: https://github.com/intel/llvm-test-suite/pull/908/commits/14b81c22ded545b309f92ffbfc550a76e4c7a7cf

vasilytric committed:
[SYCL][ESIMD] Add comments about used test coverage variables

vasilytric committed:
[SYCL][ESIMD] Use an iterator to avoid modification of reference data

vasilytric committed:
[SYCL][ESIMD] Update error messages

vasilytric committed:
[SYCL][ESIMD] Update comment and variables names

- Replace "indexess" with indexes
- Replace "selected_indexses_ptr" with "next_selected_index"

vasilytric committed:
[SYCL][ESIMD] Add static_assert

They will verify correctness of chosen offsets

vasilytric commented:
@v-klochkov Would you mind to review this PR?

v-klochkov reviewed (commented):
Looks good to me.

Can you please add one more separate test that would check zero and/or negative 'offset' and 'stride' cases?

vasilytric commented:
> Can you please add one more separate test that would check zero and/or negative 'offset' and 'stride' cases?

@v-klochkov  `Stride` should be equal or greater than zero
```
include\sycl/ext/intel/esimd/detail/region.hpp:41:3: error: static_assert failed due to requirement 'Size_x > 0 && Stride_x >= 0' "illegal region in x-dimension"
  static_assert(Size_x > 0 && Stride_x >= 0, "illegal region in x-dimension");
  ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

In case `Offset` is negative behavior is not clear (additional discussion is required) and I suggest to update coverage for simd 2d select in separate PR. 
@v-klochkov what do you think about this?

yuriykoch commented:
> Can you please add one more separate test that would check zero and/or negative 'offset' and 'stride' cases?

One more point. Is it OK to have separate PR for this test? We would have tests from this PR in CI with no delay.
